### ISSUE TITLE
Feature Request

### ISSUE THREAD

**Issue by MDE186** [gnosygnu/xowa#72]
Hey,

Would it be posible for the featured articles/news section to be updated when the main page is opened.

Something like it pulls the data directly from the featured articles/news section of the main wikipedia site?

Regards

**Comment by gnosygnu** [comment]
Hi! I renamed the issue title just for clarity's sake. Hope that works with you.

I have plans for an incremental update in the future. Basically, the user would be able to pick a page and then click a button to download the latest page from the online version.

It's not a near-term item though. I'd have to add some support for calling the Wikimedia API (to get the latest version) and then recursively call all templates to see if anything changed. So, it's not simple.

With that said, a few questions:
* Are you primarily interested in the Main Page? Would syncing other pages be useful?
* How often would you be planning to sync the Main Page? Every day? Whenever you're connected?
* I assume you only want this for the desktop? (and not Android?) 
* Are you building the wikis through Import Online / Import Offline (wikitext database)? Or are you downloading them through Download Central? (HTML database)

Just trying to get an idea of usage patterns before I start anything. Hope the above is clear. Thanks!

**Comment by MDE186** [comment]
Are you primarily interested in the Main Page? Would syncing other pages be useful?
- At the moment the Main Page is sufficient for my needs i'm not sure if I would use the feature on every page but I suppose it could be useful.
 
How often would you be planning to sync the Main Page? Every day? Whenever you're connected?
- I was hoping that the main page would sync automatically on first visit and cache the result thus allowing other users to view the cached content instead of loading the page on every visit. 
But to answer yes only once a day.

I assume you only want this for the desktop? (and not Android?)
- I only host the PC version on a ubuntu server. I have no plans for the android version.

Are you building the wikis through Import Online / Import Offline (wikitext database)? Or are you downloading them through Download Central? (HTML database)
- I am at present downloading the wikis from this link [http://xowa.org/home/wiki/Dashboard/Image_databases.html]. I have not tried the download central as of yet.

**Comment by gnosygnu** [comment]
Cool. Thanks for the answers. One other question: Is the ubuntu server for personal use, or are there multiple users? If the latter, do you want the page to be refreshed by admin only? Or whenever someone visits?

**Comment by MDE186** [comment]
The ubuntu server is pretty much headless. It's running xowa http server so
multiple users can access the information.

So whomever visits the page first then automatically loads the information
from Wikipedia. Then other users can just view the cached content for the
day.

**Comment by gnosygnu** [comment]
Ok. Makes sense.

I looked at the API a little today, and it requires some legwork on my side. I'll spend some time on it next week and post again here with progress. Feel free to ping me if you have questions or want an update.

Thanks.

**Comment by gnosygnu** [comment]
I added the calls to the Wikimedia API this week. For next week, I'm going to try to get a list of all related pages for the update.

**Comment by MDE186** [comment]
Awesome!! Thanks so much

**Comment by gnosygnu** [comment]
Just an update. I've been pretty busy with changes to the English Wikipedia HTML dumps recently, so haven't made any progress. However this week, I fleshed out the basic system, and have a primitive version working on my machine. See the screenshot below (it's from desktop, but HTTP server will work the same way) and compare with https://en.wikipedia.org/wiki/Main_Page.

I still need to polish things up (like a smart-update feature), but this should be part of the 2016-09-05 release. I'll comment again  in this ticket after I release it.

Thanks.

![update_html](https://cloud.githubusercontent.com/assets/4994104/18030937/6e127e3c-6c98-11e6-9019-b49b8b153180.png)

**Comment by MDE186** [comment]
Awesome work. Thanks :)

**Comment by gnosygnu** [comment]
Tonight's release has an early preview version of the feature. If you want to give it a try now, please do the following:

* Download the latest release at https://github.com/gnosygnu/xowa/releases/tag/v3.8.5.1 .
* Open up C:\xowa\bin\any\xowa\cfg\app\xowa.gfs
* Ctrl+F for "div_admin"
* Remove the HTML comments in the lines underneath. Basically, you want to go from this:
```
<!--
  <div id="xowa-portal-admin" class="portal">    
    <h3>~{<>app.user.msgs.get('xowa-portal-admin');<>}</h3>
    <div class="body">
      <ul>
        <li><a href="/wiki/Special:XowaSyncHtml?page=~{page_url}" class='xowa-hover-off'>~{<>app.user.msgs.get('xowa-portal-update_html');<>}</a></li>
      </ul>
    </div>
  </div>
-->
```
to this:
```
  <div id="xowa-portal-admin" class="portal">    
    <h3>~{<>app.user.msgs.get('xowa-portal-admin');<>}</h3>
    <div class="body">
      <ul>
        <li><a href="/wiki/Special:XowaSyncHtml?page=~{page_url}" class='xowa-hover-off'>~{<>app.user.msgs.get('xowa-portal-update_html');<>}</a></li>
      </ul>
    </div>
  </div>
```
* Backup the following file: C:\xowa\wiki\en.wikipedia.org\en.wikipedia.org-core.xowa
* Launch XOWA. Notice the "Update HTML" link on the left-hand side
* Go to en.wikipedia.org/wiki/Main_Page and click the "Update HTML". This should bring down the latest version

Some other notes:
* There are a few bugs I need to work on for next week:
    * Images don't show the first time unless the page is refreshed. I think I fixed this today, but honestly didn't get a chance to test.
    * Some images don't show. This has to do with strange Wikimedia logic that double-hosts the file on en.wikipedia.org and commons.wikimedia.org
* I still need to add an automatic way of checking for updates (i.e.: so you don't always have to push "Update HTML")
* When the final version is released, you should probably do the following:
    * Rollback to your copy of C:\xowa\wiki\en.wikipedia.org\en.wikipedia.org-core.xowa
    * Delete C:\xowa\wiki\en.wikipedia.org\en.wikipedia.org-html.user.xowa

Let me know if you find any bugs or if you have any suggestions.

Thanks.

**Comment by MDE186** [comment]
Thats awesome work.

Works well.

On the bug side:
1. The images are loading.
2. On today's update, I did see that the two image doesn't show. I'm sure
this is related to the "..strange Wikimedia logic.." you described.

On the question side.
1. If one updates the page and there's a new link which is not contained in
the offline version, will the link still work or will it throw an error
that the page doesn't exist?

Regards

**Comment by gnosygnu** [comment]
Cool! Thanks for the quick feedback! 

> 1. The images are loading.

Yeah, they were broken from desktop. I think they'd work fine for HTTP server though (which I assume is what you tried?)

> 2. On today's update, I did see that the two image doesn't show. I'm sure this is related to the "..strange Wikimedia logic.." you described.

I think so. I updated now and it looks like "Yavuz Sultan Selim Bridge under construction" and "Capitoline Brutus" would be missing.

If you're interested, the explanation is that 99.99% of files are in commons.wikimedia.org . However, it looks like Wikipedia promotes Main Page images by duplicating them in en.wikipedia.org , and then pointing the Main Page at the enwiki one. Unfortunately, XOWA is still looking at commons. (BTW: XOWA does look at enwiki for the other .01% of images. Just the duplicated ones cause it problems.)

I think I have a fix in place now, but need to review. It will definitely be fixed by Sunday / Monday. Let me know if you want it earlier, and I'll throw up a debug build for you.

> If one updates the page and there's a new link which is not contained in the offline version, will the link still work or will it throw an error that the page doesn't exist?

Good question. :) I am planning to put in support for this behavior, but not for the next release. Hopefully for one soon after that.

Basically, it would be nice to have XOWA start with an empty wiki, and then grab pages incrementally as you visit them. My main concerns are that it's sort of antithetical to how XOWA works today (take the full dump at once, not incrementally). Also, some things are difficult to pull incrementally (categories, language links, etc.)

That said, I may just do a basic version that just grabs the HTML and all images. It should suffice for 99% of user needs, and I'll fill in the rest at another time.

Hope this helps.

**Comment by gnosygnu** [comment]
Another update. I'm pushing the release for a week due to other reasons. I don't have the auto-update working perfectly, but I think I have the image-fix done. Let me know if you want a debug copy earlier, or if another week is fine. Thanks.

**Comment by MDE186** [comment]
Hey,

Thanks for all the effort. I'm happy to wait another week. :)

**Comment by markusmaresch** [comment]
Hi, glad to wait. No time during this week.
:-)

**Comment by gnosygnu** [comment]
Tonight's release (https://github.com/gnosygnu/xowa/releases/tag/v3.9.2.1) adds a little more functionality for the offline-sync:

* It fixes the issue with missing images b/c of duplication across two repos (cited earlier in the thread)
* It adds an Options page (home/wiki/Options/Page_sync) and a Help page (home/wiki/App/Xtn/Page_sync)
* It adds an automatic sync option (check the Options page)

However, it still lacks a few critical features:
* Math and other "images" are not sync'd. That means that sync'ing a Math page (like https://en.wikipedia.org/wiki/Pythagorean_theorem) will lose all Math equations.
* New pages still don't sync
* There is no revert option

I'll add these over the next few releases, but hopefully this is enough to replace the existing version.

Let me know if you run into any issues.

Thanks.

**Comment by gnosygnu** [comment]
[Tonight's release](https://github.com/gnosygnu/xowa/releases/tag/v3.9.4.1) adds more support for syncing other images. Namely:
* Math
* Video thumbnails
* PDF thumbnails.

New pages are still not supported, and the revert feature is also not present.

I'm planning on working on a few more image types (buttons, scores) this week. New pages is going to be a bit hard, so I'm not sure if I'll be able to cover them. The revert is also time-consuming as it involves a new page as well as other data structures.

Let me know how important the latter two items are. Otherwise, I'd like to table them for a future release.

Thanks.

**Comment by AlexanderMisel** [comment]
I tried page syncing on v3.11.2.1 win64 but it only syncs the HTML, but I want the wikitext of the page and transcluded pages also synced.

**Comment by gnosygnu** [comment]
Unfortunately this is a known limitation. Page sync only handles HTML not wikitext.

There is a good deal of complication with syncing wikitext. I listed them in https://github.com/gnosygnu/xowa/issues/97#issuecomment-259228972 and I excerpt again below.

I can look into syncing just the wikitext for the current page as that should be relatively straightforward. However, I don't know how useful this would be without syncing the transcluded template pages. If the wikitext dump is very old (say 2016-01), then page-syncing current wikitext (say 2016-11-24) could leave you with an unworkable page.

Let me know if you have other thoughts.

Thanks.

----

- Then, XOWA would have to get all Templates and Modules that are used by Earth
  - Note that Templates and Modules are deeply nested in their calls so a lot of pages will need to be checked / updated. As an example of what's involved, scroll down to the Template / Module list in https://en.wikipedia.org/w/index.php?title=Earth&action=edit
  - Also, note that once a Template is updated, other pages calling that Template can be affected. This can cause a pretty large loop where pages using that Template should also be updated (or risk being broken). For example:
    - Template:Solar_System requires a new argument as of 2016-10-10
    - On the online site, Mercury, Venus, Mars, etc. were all edited on 2016-10-10 to add the new argument
    - The offline versions of Mercury, Venus, etc. still use the old style and would be broken after bringing down the latest Template:Solar_System.
- In addition, XOWA would probably also have to get relevant Wikidata entries. This is actually complicated as Wikidata can be called from Module pages, and there's no easy way to identify dependencies (i.e.: Earth depends on Q1234 because a line of Lua code in Module:Something calls it.)
